Multithreaded key/exit message generation

Forwarded from
- https://github.com/ethereum/staking-deposit-cli/issues/364

As suggested by Preston, an investigation into building a multithreaded solution for generating keys along with exit messages could drastically improve performance when working with large quantities.

It seems like there is some kind of race issue after adding *multiprocess* support in our `test_exit_transaction_menmonic_multiple` test. This is what [the CI outputs](https://github.com/remyroy/ethstaker-deposit-cli/actions/runs/9019969298/job/24784538394) in a non-deterministic way (sometime it works, sometime it doesn't):
```
2024-05-09T16:08:06.3531967Z ##[group]Run pytest tests
2024-05-09T16:08:06.3532415Z [36;1mpytest tests[0m
2024-05-09T16:08:06.3586034Z shell: /usr/bin/bash -e {0}
2024-05-09T16:08:06.3586444Z env:
2024-05-09T16:08:06.3586924Z   pythonLocation: /opt/hostedtoolcache/Python/3.10.14/x64
2024-05-09T16:08:06.3587572Z   PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.10.14/x64/lib/pkgconfig
2024-05-09T16:08:06.3588252Z   Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.14/x64
2024-05-09T16:08:06.3588899Z   Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.14/x64
2024-05-09T16:08:06.3589489Z   Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.14/x64
2024-05-09T16:08:06.3590163Z   LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.14/x64/lib
2024-05-09T16:08:06.3590621Z ##[endgroup]
2024-05-09T16:08:08.2956600Z ============================= test session starts ==============================
2024-05-09T16:08:08.2958120Z platform linux -- Python 3.10.14, pytest-8.1.1, pluggy-1.4.0
2024-05-09T16:08:08.2959833Z rootdir: /home/runner/work/ethstaker-deposit-cli/ethstaker-deposit-cli
2024-05-09T16:08:08.2961158Z plugins: asyncio-0.23.6
2024-05-09T16:08:08.2962352Z asyncio: mode=strict
2024-05-09T16:08:08.2962952Z collected 711 items
2024-05-09T16:08:08.2963351Z 
2024-05-09T16:08:31.6307244Z tests/test_cli/test_existing_mnemonic.py .....                           [  0%]
2024-05-09T16:08:34.7508379Z tests/test_cli/test_exit_transaction_keystore.py ....                    [  1%]
2024-05-09T16:08:36.4061359Z tests/test_cli/test_exit_transaction_mnemonic.py .F.....                 [  2%]
2024-05-09T16:08:39.2593198Z tests/test_cli/test_generate_bls_to_execution_change.py ...              [  2%]
2024-05-09T16:09:03.8446919Z tests/test_cli/test_new_mnemonic.py .......                              [  3%]
2024-05-09T16:09:08.1184764Z tests/test_cli/test_regeneration.py .                                    [  3%]
2024-05-09T16:09:08.1195770Z tests/test_credentials.py .                                              [  3%]
2024-05-09T16:09:16.3293731Z tests/test_deposit.py ........                                           [  5%]
2024-05-09T16:09:16.4031321Z tests/test_intl/test_json_schema.py .................................... [ 10%]
2024-05-09T16:09:16.4752560Z ....................................                                     [ 15%]
2024-05-09T16:09:16.5074869Z tests/test_key_handling/test_key_derivation/test_mnemonic.py ........... [ 16%]
2024-05-09T16:09:16.7078626Z ........................................................................ [ 26%]
2024-05-09T16:09:16.9109373Z ........................................................................ [ 36%]
2024-05-09T16:09:19.6588158Z ........................................................................ [ 47%]
2024-05-09T16:09:23.0039761Z ........................................................................ [ 57%]
2024-05-09T16:09:26.3645323Z ........................................................................ [ 67%]
2024-05-09T16:09:29.7298447Z ........................................................................ [ 77%]
2024-05-09T16:09:32.7004665Z ................................................................         [ 86%]
2024-05-09T16:09:32.7586341Z tests/test_key_handling/test_key_derivation/test_path.py ........        [ 87%]
2024-05-09T16:09:32.7728990Z tests/test_key_handling/test_key_derivation/test_tree.py ............... [ 89%]
2024-05-09T16:09:32.8515167Z ........                                                                 [ 90%]
2024-05-09T16:09:37.3657005Z tests/test_key_handling/test_keystore.py ............                    [ 92%]
2024-05-09T16:09:37.3675366Z tests/test_utils/test_constants.py ..                                    [ 92%]
2024-05-09T16:09:37.9183379Z tests/test_utils/test_crypto.py ..........                               [ 94%]
2024-05-09T16:09:37.9292972Z tests/test_utils/test_intl.py ..........                                 [ 95%]
2024-05-09T16:09:37.9383465Z tests/test_utils/test_ssz.py .........                                   [ 96%]
2024-05-09T16:09:40.6441651Z tests/test_utils/test_validation.py ......................               [100%]
2024-05-09T16:09:40.6443173Z 
2024-05-09T16:09:40.6443987Z =================================== FAILURES ===================================
2024-05-09T16:09:40.6445580Z ___________________ test_exit_transaction_menmonic_multiple ____________________
2024-05-09T16:09:40.6446349Z 
2024-05-09T16:09:40.6449610Z     def test_exit_transaction_menmonic_multiple() -> None:
2024-05-09T16:09:40.6450486Z         # Prepare folder
2024-05-09T16:09:40.6451457Z         my_folder_path = os.path.join(os.getcwd(), 'TESTING_TEMP_FOLDER')
2024-05-09T16:09:40.6452678Z         clean_exit_transaction_folder(my_folder_path)
2024-05-09T16:09:40.6453539Z         if not os.path.exists(my_folder_path):
2024-05-09T16:09:40.6454459Z             os.mkdir(my_folder_path)
2024-05-09T16:09:40.6455167Z     
2024-05-09T16:09:40.6455778Z         runner = CliRunner()
2024-05-09T16:09:40.6456389Z         inputs = []
2024-05-09T16:09:40.6456918Z         data = '\n'.join(inputs)
2024-05-09T16:09:40.6457677Z         arguments = [
2024-05-09T16:09:40.6458358Z             '--language', 'english',
2024-05-09T16:09:40.6459065Z             '--non_interactive',
2024-05-09T16:09:40.6459976Z             'exit-transaction-mnemonic',
2024-05-09T16:09:40.6460858Z             '--output_folder', my_folder_path,
2024-05-09T16:09:40.6461742Z             '--chain', 'mainnet',
2024-05-09T16:09:40.6463056Z             '--mnemonic', 'aban aban aban aban aban aban aban aban aban aban aban abou',
2024-05-09T16:09:40.6464318Z             '--validator_start_index', '0',
2024-05-09T16:09:40.6465315Z             '--validator_indices', '0 1 2 3',
2024-05-09T16:09:40.6466049Z             '--epoch', '1234',
2024-05-09T16:09:40.6466468Z         ]
2024-05-09T16:09:40.6466982Z         result = runner.invoke(cli, arguments, input=data)
2024-05-09T16:09:40.6467458Z     
2024-05-09T16:09:40.6467871Z >       assert result.exit_code == 0
2024-05-09T16:09:40.6468302Z E       AssertionError: assert 1 == 0
2024-05-09T16:09:40.6469046Z E        +  where 1 = <Result FileExistsError(17, 'File exists')>.exit_code
2024-05-09T16:09:40.6469661Z 
2024-05-09T16:09:40.6470047Z tests/test_cli/test_exit_transaction_mnemonic.py:80: AssertionError
2024-05-09T16:09:40.6470698Z =========================== short test summary info ============================
2024-05-09T16:09:40.6471710Z FAILED tests/test_cli/test_exit_transaction_mnemonic.py::test_exit_transaction_menmonic_multiple - AssertionError: assert 1 == 0
2024-05-09T16:09:40.6472741Z  +  where 1 = <Result FileExistsError(17, 'File exists')>.exit_code
2024-05-09T16:09:40.6473382Z =================== 1 failed, 710 passed in 94.01s (0:01:34) ===================
2024-05-09T16:09:40.7114730Z ##[error]Process completed with exit code 1.
```

This needs further test and improvements.

--- a/staking_deposit/cli/exit_transaction_keystore.py
+++ b/staking_deposit/cli/exit_transaction_keystore.py
@@ -114,6 +114,8 @@ def exit_transaction_keystore(
     )
 
     folder = os.path.join(output_folder, DEFAULT_EXIT_TRANSACTION_FOLDER_NAME)
+    if not os.path.exists(folder):
+        os.mkdir(folder)
 
     click.echo(load_text(['msg_exit_transaction_creation']))
     saved_folder = export_exit_transaction_json(folder=folder, signed_exit=signed_exit)

--- a/staking_deposit/cli/exit_transaction_mnemonic.py
+++ b/staking_deposit/cli/exit_transaction_mnemonic.py
@@ -1,7 +1,8 @@
 import click
 import os
+import concurrent.futures
 
-from typing import Any, Sequence
+from typing import Any, Sequence, Dict
 from staking_deposit.cli.existing_mnemonic import load_mnemonic_arguments_decorator
 from staking_deposit.credentials import Credential
 from staking_deposit.exceptions import ValidationError
@@ -21,6 +22,22 @@ from staking_deposit.utils.intl import (
     load_text,
 )
 from staking_deposit.utils.validation import validate_int_range, validate_validator_indices, verify_signed_exit_json
+
+
+def _credential_builder(kwargs: Dict[str, Any]) -> Credential:
+    return Credential(**kwargs)
+
+
+def _exit_exporter(kwargs: Dict[str, Any]) -> str:
+    credential: Credential = kwargs.pop('credential')
+    return credential.save_exit_transaction(**kwargs)
+
+
+def _exit_verifier(kwargs: Dict[str, Any]) -> bool:
+    credential: Credential = kwargs.pop('credential')
+    kwargs['pubkey'] = credential.signing_pk.hex()
+    kwargs['chain_settings'] = credential.chain_setting
+    return verify_signed_exit_json(**kwargs)
 
 
 FUNC_NAME = 'exit_transaction_mnemonic'
@@ -94,39 +111,56 @@ def exit_transaction_mnemonic(
     key_indices = range(validator_start_index, validator_start_index + num_keys)
 
     # We are not using CredentialList because from_mnemonic assumes key generation flow
-    credentials = [
-        Credential(
-            mnemonic=mnemonic,
-            mnemonic_password=mnemonic_password,
-            index=key_index,
-            amount=0,  # Unneeded for this purpose
-            chain_setting=chain_settings,
-            hex_eth1_withdrawal_address=None
-        ) for key_index in key_indices
-    ]
+    credentials = []
+    with click.progressbar(length=num_keys, label=load_text(['msg_key_creation']),
+                           show_percent=False, show_pos=True) as bar:
 
-    with click.progressbar(zip(credentials, validator_indices),
-                           label=load_text(['msg_exit_transaction_creation']),
-                           show_percent=False,
-                           length=num_keys,
-                           show_pos=True) as items:
-        transaction_filefolders = [
-            credential.save_exit_transaction(validator_index=validator_index, epoch=epoch, folder=folder)
-            for credential, validator_index in items
-        ]
+        executor_kwargs = [{
+            'mnemonic': mnemonic,
+            'mnemonic_password': mnemonic_password,
+            'index': index,
+            'amount': 0,
+            'chain_setting': chain_settings,
+            'hex_eth1_withdrawal_address': None,
+        } for index in key_indices]
 
-    with click.progressbar(zip(transaction_filefolders, credentials),
-                           label=load_text(['msg_verify_exit_transaction']),
-                           show_percent=False,
-                           length=num_keys,
-                           show_pos=True) as items:
-        if not all(
-            verify_signed_exit_json(file_folder=file,
-                                    pubkey=credential.signing_pk.hex(),
-                                    chain_settings=credential.chain_setting)
-            for file, credential in items
-        ):
-            raise ValidationError(load_text(['err_verify_exit_transactions']))
+        with concurrent.futures.ProcessPoolExecutor() as executor:
+            for credential in executor.map(_credential_builder, executor_kwargs):
+                credentials.append(credential)
+                bar.update(1)
+
+    transaction_filefolders = []
+    with click.progressbar(length=num_keys, label=load_text(['msg_exit_transaction_creation']),
+                           show_percent=False, show_pos=True) as bar:
+
+        executor_kwargs = [{
+            'credential': credential,
+            'validator_index': validator_index,
+            'epoch': epoch,
+            'folder': folder,
+        } for credential, validator_index in zip(credentials, validator_indices)]
+
+        with concurrent.futures.ProcessPoolExecutor() as executor:
+            for filefolder in executor.map(_exit_exporter, executor_kwargs):
+                transaction_filefolders.append(filefolder)
+                bar.update(1)
+
+    all_valid_exits = True
+    with click.progressbar(length=num_keys, label=load_text(['msg_verify_exit_transaction']),
+                           show_percent=False, show_pos=True) as bar:
+
+        executor_kwargs = [{
+            'file_folder': file,
+            'credential': credential,
+        } for file, credential in zip(transaction_filefolders, credentials)]
+
+        with concurrent.futures.ProcessPoolExecutor() as executor:
+            for valid_exit in executor.map(_exit_verifier, executor_kwargs):
+                all_valid_exits &= valid_exit
+                bar.update(1)
+
+    if not all_valid_exits:
+        raise ValidationError(load_text(['err_verify_exit_transactions']))
 
     click.echo(load_text(['msg_creation_success']) + folder)
     click.pause(load_text(['msg_pause']))

--- a/staking_deposit/cli/exit_transaction_mnemonic.py
+++ b/staking_deposit/cli/exit_transaction_mnemonic.py
@@ -129,6 +129,9 @@ def exit_transaction_mnemonic(
                 credentials.append(credential)
                 bar.update(1)
 
+    if not os.path.exists(folder):
+        os.mkdir(folder)
+
     transaction_filefolders = []
     with click.progressbar(length=num_keys, label=load_text(['msg_exit_transaction_creation']),
                            show_percent=False, show_pos=True) as bar:

--- a/staking_deposit/cli/exit_transaction_mnemonic.py
+++ b/staking_deposit/cli/exit_transaction_mnemonic.py
@@ -145,7 +145,6 @@ def exit_transaction_mnemonic(
                 transaction_filefolders.append(filefolder)
                 bar.update(1)
 
-    all_valid_exits = True
     with click.progressbar(length=num_keys, label=load_text(['msg_verify_exit_transaction']),
                            show_percent=False, show_pos=True) as bar:
 
@@ -156,11 +155,9 @@ def exit_transaction_mnemonic(
 
         with concurrent.futures.ProcessPoolExecutor() as executor:
             for valid_exit in executor.map(_exit_verifier, executor_kwargs):
-                all_valid_exits &= valid_exit
                 bar.update(1)
-
-    if not all_valid_exits:
-        raise ValidationError(load_text(['err_verify_exit_transactions']))
+                if not valid_exit:
+                    raise ValidationError(load_text(['err_verify_exit_transactions']))
 
     click.echo(load_text(['msg_creation_success']) + folder)
     click.pause(load_text(['msg_pause']))

--- a/staking_deposit/cli/generate_bls_to_execution_change.py
+++ b/staking_deposit/cli/generate_bls_to_execution_change.py
@@ -203,7 +203,7 @@ def generate_bls_to_execution_change(
             for e in executor.map(_validate_credentials_match, executor_kwargs):
                 bar.update(1)
                 if e is not None:
-                    click.echo('\n[Error] ' + str(e))
+                    click.echo('\n\n[Error] ' + str(e))
                     return
 
     btec_file = credentials.export_bls_to_execution_change_json(bls_to_execution_changes_folder, validator_indices)

--- a/staking_deposit/cli/generate_bls_to_execution_change.py
+++ b/staking_deposit/cli/generate_bls_to_execution_change.py
@@ -1,15 +1,19 @@
 import os
 import click
 import json
+import concurrent.futures
 from typing import (
     Any,
     Sequence,
+    Dict,
+    Optional
 )
 
 from eth_typing import HexAddress
 
 from staking_deposit.credentials import (
     CredentialList,
+    Credential
 )
 from staking_deposit.utils.validation import (
     validate_bls_withdrawal_credentials_list,
@@ -46,6 +50,17 @@ from .existing_mnemonic import (
 
 def get_password(text: str) -> str:
     return click.prompt(text, hide_input=True, show_default=False, type=str)
+
+
+def _validate_credentials_match(kwargs: Dict[str, Any]) -> Optional[ValidationError]:
+    credential: Credential = kwargs.pop('credential')
+    bls_withdrawal_credentials: bytes = kwargs.pop('bls_withdrawal_credentials')
+
+    try:
+        validate_bls_withdrawal_credentials_matching(bls_withdrawal_credentials, credential)
+    except ValidationError as e:
+        return e
+    return None
 
 
 FUNC_NAME = 'generate_bls_to_execution_change'
@@ -177,12 +192,19 @@ def generate_bls_to_execution_change(
     )
 
     # Check if the given old bls_withdrawal_credentials is as same as the mnemonic generated
-    for i, credential in enumerate(credentials.credentials):
-        try:
-            validate_bls_withdrawal_credentials_matching(bls_withdrawal_credentials_list[i], credential)
-        except ValidationError as e:
-            click.echo('\n[Error] ' + str(e))
-            return
+    with click.progressbar(length=len(credentials.credentials), label=load_text(['msg_credentials_verification']),
+                           show_percent=False, show_pos=True) as bar:
+        executor_kwargs = [{
+            'credential': credential,
+            'bls_withdrawal_credentials': bls_withdrawal_credentials_list[i],
+        } for i, credential in enumerate(credentials.credentials)]
+
+        with concurrent.futures.ProcessPoolExecutor() as executor:
+            for e in executor.map(_validate_credentials_match, executor_kwargs):
+                bar.update(1)
+                if e is not None:
+                    click.echo('\n[Error] ' + str(e))
+                    return
 
     btec_file = credentials.export_bls_to_execution_change_json(bls_to_execution_changes_folder, validator_indices)
 

--- a/staking_deposit/credentials.py
+++ b/staking_deposit/credentials.py
@@ -243,6 +243,11 @@ def _keystore_verifier(kwargs: Dict[str, Any]) -> bool:
     return credential.verify_keystore(**kwargs)
 
 
+def _bls_to_execution_change_builder(kwargs: Dict[str, Any]) -> Dict[str, bytes]:
+    credential: Credential = kwargs.pop('credential')
+    return credential.get_bls_to_execution_change_dict(**kwargs)
+
+
 class CredentialList:
     """
     A collection of multiple Credentials, one for each validator.
@@ -269,7 +274,7 @@ class CredentialList:
         return_list: List[Credential] = []
         with click.progressbar(length=num_keys, label=load_text(['msg_key_creation']),
                                show_percent=False, show_pos=True) as bar:
-            kwargs = [{
+            executor_kwargs = [{
                 'mnemonic': mnemonic,
                 'mnemonic_password': mnemonic_password,
                 'index': index,
@@ -279,7 +284,7 @@ class CredentialList:
             } for index in key_indices]
 
             with concurrent.futures.ProcessPoolExecutor() as executor:
-                for credential in executor.map(_credential_builder, kwargs):
+                for credential in executor.map(_credential_builder, executor_kwargs):
                     return_list.append(credential)
                     bar.update(1)
         return cls(return_list)
@@ -288,14 +293,14 @@ class CredentialList:
         return_list: List[str] = []
         with click.progressbar(length=len(self.credentials), label=load_text(['msg_keystore_creation']),
                                show_percent=False, show_pos=True) as bar:
-            kwargs = [{
+            executor_kwargs = [{
                 'credential': credential,
                 'password': password,
                 'folder': folder,
             } for credential in self.credentials]
 
             with concurrent.futures.ProcessPoolExecutor() as executor:
-                for keystore in executor.map(_keystore_exporter, kwargs):
+                for keystore in executor.map(_keystore_exporter, executor_kwargs):
                     return_list.append(keystore)
                     bar.update(1)
         return return_list
@@ -322,24 +327,33 @@ class CredentialList:
         with click.progressbar(length=len(self.credentials),
                                label=load_text(['msg_keystore_verification']),
                                show_percent=False, show_pos=True) as bar:
-            kwargs = [{
+            executor_kwargs = [{
                 'credential': credential,
                 'keystore_filefolder': fileholder,
                 'password': password,
             } for credential, fileholder in zip(self.credentials, keystore_filefolders)]
 
             with concurrent.futures.ProcessPoolExecutor() as executor:
-                for valid_keystore in executor.map(_keystore_verifier, kwargs):
+                for valid_keystore in executor.map(_keystore_verifier, executor_kwargs):
                     valid &= valid_keystore
                     bar.update(1)
 
         return valid
 
     def export_bls_to_execution_change_json(self, folder: str, validator_indices: Sequence[int]) -> str:
-        with click.progressbar(self.credentials, label=load_text(['msg_bls_to_execution_change_creation']),
-                               show_percent=False, show_pos=True) as credentials:
-            bls_to_execution_changes = [cred.get_bls_to_execution_change_dict(validator_indices[i])
-                                        for i, cred in enumerate(credentials)]
+        bls_to_execution_changes = []
+        with click.progressbar(length=len(self.credentials), label=load_text(['msg_bls_to_execution_change_creation']),
+                               show_percent=False, show_pos=True) as bar:
+
+            executor_kwargs = [{
+                'credential': credential,
+                'validator_index': validator_indices[i],
+            } for i, credential in enumerate(self.credentials)]
+
+            with concurrent.futures.ProcessPoolExecutor() as executor:
+                for bls_to_execution_change in executor.map(_bls_to_execution_change_builder, executor_kwargs):
+                    bls_to_execution_changes.append(bls_to_execution_change)
+                    bar.update(1)
 
         filefolder = os.path.join(folder, 'bls_to_execution_change-%i.json' % time.time())
         with open(filefolder, 'w') as f:

--- a/staking_deposit/credentials.py
+++ b/staking_deposit/credentials.py
@@ -4,6 +4,7 @@ from enum import Enum
 import time
 import json
 from typing import Dict, List, Optional, Any, Sequence
+import concurrent.futures
 
 from eth_typing import Address, HexAddress
 from eth_utils import to_canonical_address
@@ -224,6 +225,15 @@ class Credential:
         return export_exit_transaction_json(folder=folder, signed_exit=signed_voluntary_exit)
 
 
+def credential_builder(kwargs: Dict[str, Any]) -> Credential:
+    return Credential(**kwargs)
+
+
+def keystore_exporter(kwargs: Dict[str, Any]) -> str:
+    credential: Credential = kwargs.pop('credential')
+    return credential.save_signing_keystore(**kwargs)
+
+
 class CredentialList:
     """
     A collection of multiple Credentials, one for each validator.
@@ -246,17 +256,40 @@ class CredentialList:
                 f"The number of keys ({num_keys}) doesn't equal to the corresponding deposit amounts ({len(amounts)})."
             )
         key_indices = range(start_index, start_index + num_keys)
-        with click.progressbar(key_indices, label=load_text(['msg_key_creation']),
-                               show_percent=False, show_pos=True) as indices:
-            return cls([Credential(mnemonic=mnemonic, mnemonic_password=mnemonic_password,
-                                   index=index, amount=amounts[index - start_index], chain_setting=chain_setting,
-                                   hex_eth1_withdrawal_address=hex_eth1_withdrawal_address)
-                        for index in indices])
+
+        return_list: List[Credential] = []
+        with click.progressbar(length=num_keys, label=load_text(['msg_key_creation']),
+                               show_percent=False, show_pos=True) as bar:
+            kwargs = [{
+                'mnemonic': mnemonic,
+                'mnemonic_password': mnemonic_password,
+                'index': index,
+                'amount': amounts[index - start_index],
+                'chain_setting': chain_setting,
+                'hex_eth1_withdrawal_address': hex_eth1_withdrawal_address,
+            } for index in key_indices]
+
+            with concurrent.futures.ProcessPoolExecutor() as executor:
+                for index, credential in zip(key_indices, executor.map(credential_builder, kwargs)):
+                    return_list.append(credential)
+                    bar.update(1)
+        return cls(return_list)
 
     def export_keystores(self, password: str, folder: str) -> List[str]:
-        with click.progressbar(self.credentials, label=load_text(['msg_keystore_creation']),
-                               show_percent=False, show_pos=True) as credentials:
-            return [credential.save_signing_keystore(password=password, folder=folder) for credential in credentials]
+        return_list: List[str] = []
+        with click.progressbar(length=len(self.credentials), label=load_text(['msg_keystore_creation']),
+                               show_percent=False, show_pos=True) as bar:
+            kwargs = [{
+                'credential': credential,
+                'password': password,
+                'folder': folder,
+            } for credential in self.credentials]
+
+            with concurrent.futures.ProcessPoolExecutor() as executor:
+                for credential, keystore in zip(self.credentials, executor.map(keystore_exporter, kwargs)):
+                    return_list.append(keystore)
+                    bar.update(1)
+        return return_list
 
     def export_deposit_data_json(self, folder: str) -> str:
         with click.progressbar(self.credentials, label=load_text(['msg_depositdata_creation']),

--- a/staking_deposit/credentials.py
+++ b/staking_deposit/credentials.py
@@ -323,7 +323,7 @@ class CredentialList:
         return filefolder
 
     def verify_keystores(self, keystore_filefolders: List[str], password: str) -> bool:
-        valid = True
+        all_valid_keystores = True
         with click.progressbar(length=len(self.credentials),
                                label=load_text(['msg_keystore_verification']),
                                show_percent=False, show_pos=True) as bar:
@@ -335,10 +335,10 @@ class CredentialList:
 
             with concurrent.futures.ProcessPoolExecutor() as executor:
                 for valid_keystore in executor.map(_keystore_verifier, executor_kwargs):
-                    valid &= valid_keystore
+                    all_valid_keystores &= valid_keystore
                     bar.update(1)
 
-        return valid
+        return all_valid_keystores
 
     def export_bls_to_execution_change_json(self, folder: str, validator_indices: Sequence[int]) -> str:
         bls_to_execution_changes = []

--- a/staking_deposit/credentials.py
+++ b/staking_deposit/credentials.py
@@ -3,8 +3,8 @@ import click
 from enum import Enum
 import time
 import json
-from typing import Dict, List, Optional, Any, Sequence
 import concurrent.futures
+from typing import Dict, List, Optional, Any, Sequence
 
 from eth_typing import Address, HexAddress
 from eth_utils import to_canonical_address
@@ -225,13 +225,22 @@ class Credential:
         return export_exit_transaction_json(folder=folder, signed_exit=signed_voluntary_exit)
 
 
-def credential_builder(kwargs: Dict[str, Any]) -> Credential:
+def _credential_builder(kwargs: Dict[str, Any]) -> Credential:
     return Credential(**kwargs)
 
 
-def keystore_exporter(kwargs: Dict[str, Any]) -> str:
+def _keystore_exporter(kwargs: Dict[str, Any]) -> str:
     credential: Credential = kwargs.pop('credential')
     return credential.save_signing_keystore(**kwargs)
+
+
+def _deposit_data_builder(credential: Credential) -> Dict[str, bytes]:
+    return credential.deposit_datum_dict
+
+
+def _keystore_verifier(kwargs: Dict[str, Any]) -> bool:
+    credential: Credential = kwargs.pop('credential')
+    return credential.verify_keystore(**kwargs)
 
 
 class CredentialList:
@@ -270,7 +279,7 @@ class CredentialList:
             } for index in key_indices]
 
             with concurrent.futures.ProcessPoolExecutor() as executor:
-                for index, credential in zip(key_indices, executor.map(credential_builder, kwargs)):
+                for credential in executor.map(_credential_builder, kwargs):
                     return_list.append(credential)
                     bar.update(1)
         return cls(return_list)
@@ -286,15 +295,21 @@ class CredentialList:
             } for credential in self.credentials]
 
             with concurrent.futures.ProcessPoolExecutor() as executor:
-                for credential, keystore in zip(self.credentials, executor.map(keystore_exporter, kwargs)):
+                for keystore in executor.map(_keystore_exporter, kwargs):
                     return_list.append(keystore)
                     bar.update(1)
         return return_list
 
     def export_deposit_data_json(self, folder: str) -> str:
-        with click.progressbar(self.credentials, label=load_text(['msg_depositdata_creation']),
-                               show_percent=False, show_pos=True) as credentials:
-            deposit_data = [cred.deposit_datum_dict for cred in credentials]
+        deposit_data = []
+        with click.progressbar(length=len(self.credentials), label=load_text(['msg_depositdata_creation']),
+                               show_percent=False, show_pos=True) as bar:
+
+            with concurrent.futures.ProcessPoolExecutor() as executor:
+                for datum_dict in executor.map(_deposit_data_builder, self.credentials):
+                    deposit_data.append(datum_dict)
+                    bar.update(1)
+
         filefolder = os.path.join(folder, 'deposit_data-%i.json' % time.time())
         with open(filefolder, 'w') as f:
             json.dump(deposit_data, f, default=lambda x: x.hex())
@@ -303,11 +318,22 @@ class CredentialList:
         return filefolder
 
     def verify_keystores(self, keystore_filefolders: List[str], password: str) -> bool:
-        with click.progressbar(zip(self.credentials, keystore_filefolders),
+        valid = True
+        with click.progressbar(length=len(self.credentials),
                                label=load_text(['msg_keystore_verification']),
-                               length=len(self.credentials), show_percent=False, show_pos=True) as items:
-            return all(credential.verify_keystore(keystore_filefolder=filefolder, password=password)
-                       for credential, filefolder in items)
+                               show_percent=False, show_pos=True) as bar:
+            kwargs = [{
+                'credential': credential,
+                'keystore_filefolder': fileholder,
+                'password': password,
+            } for credential, fileholder in zip(self.credentials, keystore_filefolders)]
+
+            with concurrent.futures.ProcessPoolExecutor() as executor:
+                for valid_keystore in executor.map(_keystore_verifier, kwargs):
+                    valid &= valid_keystore
+                    bar.update(1)
+
+        return valid
 
     def export_bls_to_execution_change_json(self, folder: str, validator_indices: Sequence[int]) -> str:
         with click.progressbar(self.credentials, label=load_text(['msg_bls_to_execution_change_creation']),

--- a/staking_deposit/credentials.py
+++ b/staking_deposit/credentials.py
@@ -271,7 +271,7 @@ class CredentialList:
             )
         key_indices = range(start_index, start_index + num_keys)
 
-        return_list: List[Credential] = []
+        credentials: List[Credential] = []
         with click.progressbar(length=num_keys, label=load_text(['msg_key_creation']),
                                show_percent=False, show_pos=True) as bar:
             executor_kwargs = [{
@@ -285,12 +285,12 @@ class CredentialList:
 
             with concurrent.futures.ProcessPoolExecutor() as executor:
                 for credential in executor.map(_credential_builder, executor_kwargs):
-                    return_list.append(credential)
+                    credentials.append(credential)
                     bar.update(1)
-        return cls(return_list)
+        return cls(credentials)
 
     def export_keystores(self, password: str, folder: str) -> List[str]:
-        return_list: List[str] = []
+        filefolders: List[str] = []
         with click.progressbar(length=len(self.credentials), label=load_text(['msg_keystore_creation']),
                                show_percent=False, show_pos=True) as bar:
             executor_kwargs = [{
@@ -300,10 +300,10 @@ class CredentialList:
             } for credential in self.credentials]
 
             with concurrent.futures.ProcessPoolExecutor() as executor:
-                for keystore in executor.map(_keystore_exporter, executor_kwargs):
-                    return_list.append(keystore)
+                for filefolder in executor.map(_keystore_exporter, executor_kwargs):
+                    filefolders.append(filefolder)
                     bar.update(1)
-        return return_list
+        return filefolders
 
     def export_deposit_data_json(self, folder: str) -> str:
         deposit_data = []

--- a/staking_deposit/deposit.py
+++ b/staking_deposit/deposit.py
@@ -92,7 +92,7 @@ cli.add_command(exit_transaction_mnemonic)
 
 
 def run() -> None:
-    freeze_support() # Needed when running under Windows in a frozen bundle
+    freeze_support()  # Needed when running under Windows in a frozen bundle
     check_python_version()
     cli()
 

--- a/staking_deposit/exit_transaction.py
+++ b/staking_deposit/exit_transaction.py
@@ -48,9 +48,6 @@ def export_exit_transaction_json(folder: str, signed_exit: SignedVoluntaryExit) 
     signed_exit_json.update({'message': message})
     signed_exit_json.update({'signature': '0x' + signed_exit.signature.hex()})  # type: ignore[attr-defined]
 
-    if not os.path.exists(folder):
-        os.mkdir(folder)
-
     filefolder = os.path.join(
         folder,
         'signed_exit_transaction-%s-%i.json' % (

--- a/staking_deposit/intl/en/cli/exit_transaction_mnemonic.json
+++ b/staking_deposit/intl/en/cli/exit_transaction_mnemonic.json
@@ -21,6 +21,7 @@
       "arg_exit_transaction_mnemonic_output_folder": {
           "help": "The folder path where the exit transactions will be saved to. Pointing to `./exit_transactions` by default."
       },
+      "msg_key_creation": "Creating your keys:\t",
       "msg_exit_transaction_creation": "Creating your exit transactions:\t",
       "msg_verify_exit_transaction": "Verifying your exit transactions:\t",
       "err_verify_exit_transactions": "\nThere was a problem verifying your exit transactions.\nPlease try again",

--- a/staking_deposit/intl/en/cli/generate_bls_to_execution_change.json
+++ b/staking_deposit/intl/en/cli/generate_bls_to_execution_change.json
@@ -34,6 +34,7 @@
             "help": "The folder path for the keystore(s). Pointing to `./bls_to_execution_changes` by default."
         },
         "msg_key_creation": "Creating your SignedBLSToExecutionChange.",
+        "msg_credentials_verification": "Verifying your withdrawal credentials.",
         "msg_creation_success": "\nSuccess!\nYour SignedBLSToExecutionChange JSON file can be found at: ",
         "msg_pause": "\n\nPress any key.",
         "err_verify_btec": "Failed to verify the bls_to_execution_change JSON files."

--- a/staking_deposit/utils/validation.py
+++ b/staking_deposit/utils/validation.py
@@ -53,7 +53,7 @@ def verify_deposit_data_json(filefolder: str, credentials: Sequence[Credential])
     """
     Validate every deposit found in the deposit-data JSON file folder.
     """
-    valid = True
+    all_valid_deposits = True
     deposit_json = []
     with open(filefolder, 'r', encoding='utf-8') as f:
         deposit_json = json.load(f)
@@ -67,10 +67,10 @@ def verify_deposit_data_json(filefolder: str, credentials: Sequence[Credential])
 
         with concurrent.futures.ProcessPoolExecutor() as executor:
             for valid_deposit in executor.map(_deposit_validator, executor_kwargs):
-                valid &= valid_deposit
+                all_valid_deposits &= valid_deposit
                 bar.update(1)
 
-    return valid
+    return all_valid_deposits
 
 
 def validate_deposit(deposit_data_dict: Dict[str, Any], credential: Credential) -> bool:
@@ -185,7 +185,7 @@ def verify_bls_to_execution_change_json(filefolder: str,
     with open(filefolder, 'r', encoding='utf-8') as f:
         btec_json = json.load(f)
 
-    valid = True
+    all_valid_bls_changes = True
     with click.progressbar(length=len(btec_json), label=load_text(['msg_bls_to_execution_change_verification']),
                            show_percent=False, show_pos=True) as bar:
 
@@ -199,10 +199,10 @@ def verify_bls_to_execution_change_json(filefolder: str,
 
         with concurrent.futures.ProcessPoolExecutor() as executor:
             for valid_bls_change in executor.map(_bls_to_execution_change_validator, executor_kwargs):
-                valid &= valid_bls_change
+                all_valid_bls_changes &= valid_bls_change
                 bar.update(1)
 
-    return valid
+    return all_valid_bls_changes
 
 
 def validate_bls_to_execution_change(btec_dict: Dict[str, Any],

--- a/staking_deposit/utils/validation.py
+++ b/staking_deposit/utils/validation.py
@@ -43,11 +43,6 @@ from staking_deposit.settings import BaseChainSetting
 # Deposit
 #
 
-def _deposit_validator(kwargs: Dict[str, Any]) -> bool:
-    deposit: Dict[str, Any] = kwargs.pop('deposit')
-    credential: Credential = kwargs.pop('credential')
-    return validate_deposit(deposit, credential)
-
 
 def verify_deposit_data_json(filefolder: str, credentials: Sequence[Credential]) -> bool:
     """
@@ -60,13 +55,9 @@ def verify_deposit_data_json(filefolder: str, credentials: Sequence[Credential])
 
     with click.progressbar(length=len(deposit_json), label=load_text(['msg_deposit_verification']),
                            show_percent=False, show_pos=True) as bar:
-        executor_kwargs = [{
-            'credential': credential,
-            'deposit': deposit,
-        } for deposit, credential in zip(deposit_json, credentials)]
 
         with concurrent.futures.ProcessPoolExecutor() as executor:
-            for valid_deposit in executor.map(_deposit_validator, executor_kwargs):
+            for valid_deposit in executor.map(validate_deposit, deposit_json, credentials):
                 all_valid_deposits &= valid_deposit
                 bar.update(1)
 

--- a/staking_deposit/utils/validation.py
+++ b/staking_deposit/utils/validation.py
@@ -1,6 +1,7 @@
 import click
 import json
 import re
+import concurrent.futures
 from typing import Any, Dict, Sequence
 
 from eth_typing import (
@@ -42,16 +43,34 @@ from staking_deposit.settings import BaseChainSetting
 # Deposit
 #
 
+def _deposit_validator(kwargs: Dict[str, Any]) -> bool:
+    deposit: Dict[str, Any] = kwargs.pop('deposit')
+    credential: Credential = kwargs.pop('credential')
+    return validate_deposit(deposit, credential)
+
+
 def verify_deposit_data_json(filefolder: str, credentials: Sequence[Credential]) -> bool:
     """
     Validate every deposit found in the deposit-data JSON file folder.
     """
+    valid = True
+    deposit_json = []
     with open(filefolder, 'r', encoding='utf-8') as f:
         deposit_json = json.load(f)
-        with click.progressbar(deposit_json, label=load_text(['msg_deposit_verification']),
-                               show_percent=False, show_pos=True) as deposits:
-            return all([validate_deposit(deposit, credential) for deposit, credential in zip(deposits, credentials)])
-    return False
+
+    with click.progressbar(length=len(deposit_json), label=load_text(['msg_deposit_verification']),
+                           show_percent=False, show_pos=True) as bar:
+        kwargs = [{
+            'credential': credential,
+            'deposit': deposit,
+        } for deposit, credential in zip(deposit_json, credentials)]
+
+        with concurrent.futures.ProcessPoolExecutor() as executor:
+            for valid_deposit in executor.map(_deposit_validator, kwargs):
+                valid &= valid_deposit
+                bar.update(1)
+
+    return valid
 
 
 def validate_deposit(deposit_data_dict: Dict[str, Any], credential: Credential) -> bool:

--- a/staking_deposit/utils/validation.py
+++ b/staking_deposit/utils/validation.py
@@ -60,13 +60,13 @@ def verify_deposit_data_json(filefolder: str, credentials: Sequence[Credential])
 
     with click.progressbar(length=len(deposit_json), label=load_text(['msg_deposit_verification']),
                            show_percent=False, show_pos=True) as bar:
-        kwargs = [{
+        executor_kwargs = [{
             'credential': credential,
             'deposit': deposit,
         } for deposit, credential in zip(deposit_json, credentials)]
 
         with concurrent.futures.ProcessPoolExecutor() as executor:
-            for valid_deposit in executor.map(_deposit_validator, kwargs):
+            for valid_deposit in executor.map(_deposit_validator, executor_kwargs):
                 valid &= valid_deposit
                 bar.update(1)
 
@@ -168,6 +168,9 @@ def validate_eth1_withdrawal_address(cts: click.Context, param: Any, address: st
 # BLSToExecutionChange
 #
 
+def _bls_to_execution_change_validator(kwargs: Dict[str, Any]) -> bool:
+    return validate_bls_to_execution_change(**kwargs)
+
 
 def verify_bls_to_execution_change_json(filefolder: str,
                                         credentials: Sequence[Credential],
@@ -178,19 +181,28 @@ def verify_bls_to_execution_change_json(filefolder: str,
     """
     Validate every BLSToExecutionChange found in the bls_to_execution_change JSON file folder.
     """
+    btec_json = []
     with open(filefolder, 'r', encoding='utf-8') as f:
         btec_json = json.load(f)
-        with click.progressbar(btec_json, label=load_text(['msg_bls_to_execution_change_verification']),
-                               show_percent=False, show_pos=True) as btecs:
-            return all([
-                validate_bls_to_execution_change(
-                    btec, credential,
-                    input_validator_index=input_validator_index,
-                    input_execution_address=input_execution_address,
-                    chain_setting=chain_setting)
-                for btec, credential, input_validator_index in zip(btecs, credentials, input_validator_indices)
-            ])
-    return False
+
+    valid = True
+    with click.progressbar(length=len(btec_json), label=load_text(['msg_bls_to_execution_change_verification']),
+                           show_percent=False, show_pos=True) as bar:
+
+        executor_kwargs = [{
+            'btec_dict': btec,
+            'credential': credential,
+            'input_validator_index': input_validator_index,
+            'input_execution_address': input_execution_address,
+            'chain_setting': chain_setting,
+        } for btec, credential, input_validator_index in zip(btec_json, credentials, input_validator_indices)]
+
+        with concurrent.futures.ProcessPoolExecutor() as executor:
+            for valid_bls_change in executor.map(_bls_to_execution_change_validator, executor_kwargs):
+                valid &= valid_bls_change
+                bar.update(1)
+
+    return valid
 
 
 def validate_bls_to_execution_change(btec_dict: Dict[str, Any],

--- a/test_binary_btec_script.py
+++ b/test_binary_btec_script.py
@@ -61,6 +61,8 @@ async def main(argv):
 
     assert len(seed_phrase) > 0
 
+    await proc.wait()
+
     # Check files
     validator_keys_folder_path = os.path.join(my_folder_path, DEFAULT_VALIDATOR_KEYS_FOLDER_NAME)
     _, _, key_files = next(os.walk(validator_keys_folder_path))

--- a/test_binary_deposit_script.py
+++ b/test_binary_deposit_script.py
@@ -57,6 +57,8 @@ async def main(argv):
 
     assert len(seed_phrase) > 0
 
+    await proc.wait()
+
     # Check files
     validator_keys_folder_path = os.path.join(my_folder_path, DEFAULT_VALIDATOR_KEYS_FOLDER_NAME)
     _, _, key_files = next(os.walk(validator_keys_folder_path))

--- a/test_btec_script.py
+++ b/test_btec_script.py
@@ -66,6 +66,8 @@ async def main():
 
     assert len(seed_phrase) > 0
 
+    await proc.wait()
+
     # Check files
     validator_keys_folder_path = os.path.join(my_folder_path, DEFAULT_VALIDATOR_KEYS_FOLDER_NAME)
     _, _, key_files = next(os.walk(validator_keys_folder_path))

--- a/test_deposit_script.py
+++ b/test_deposit_script.py
@@ -62,6 +62,8 @@ async def main():
 
     assert len(seed_phrase) > 0
 
+    await proc.wait()
+
     # Check files
     validator_keys_folder_path = os.path.join(my_folder_path, DEFAULT_VALIDATOR_KEYS_FOLDER_NAME)
     _, _, key_files = next(os.walk(validator_keys_folder_path))

--- a/tests/test_cli/test_existing_mnemonic.py
+++ b/tests/test_cli/test_existing_mnemonic.py
@@ -281,6 +281,7 @@ async def test_script() -> None:
         ' '.join(cmd_args),
     )
     await proc.wait()
+
     # Check files
     validator_keys_folder_path = os.path.join(my_folder_path, DEFAULT_VALIDATOR_KEYS_FOLDER_NAME)
     _, _, key_files = next(os.walk(validator_keys_folder_path))
@@ -329,6 +330,7 @@ async def test_script_abbreviated_mnemonic() -> None:
         ' '.join(cmd_args),
     )
     await proc.wait()
+
     # Check files
     validator_keys_folder_path = os.path.join(my_folder_path, DEFAULT_VALIDATOR_KEYS_FOLDER_NAME)
     _, _, key_files = next(os.walk(validator_keys_folder_path))

--- a/tests/test_cli/test_exit_transaction_mnemonic.py
+++ b/tests/test_cli/test_exit_transaction_mnemonic.py
@@ -1,3 +1,4 @@
+import asyncio
 import os
 import pytest
 
@@ -9,7 +10,7 @@ from staking_deposit.utils.constants import DEFAULT_EXIT_TRANSACTION_FOLDER_NAME
 from tests.test_cli.helpers import clean_exit_transaction_folder, read_json_file, verify_file_permission
 
 
-def test_exit_transaction_menmonic() -> None:
+def test_exit_transaction_mnemonic() -> None:
     # Prepare folder
     my_folder_path = os.path.join(os.getcwd(), 'TESTING_TEMP_FOLDER')
     clean_exit_transaction_folder(my_folder_path)
@@ -54,30 +55,43 @@ def test_exit_transaction_menmonic() -> None:
     clean_exit_transaction_folder(my_folder_path)
 
 
-def test_exit_transaction_menmonic_multiple() -> None:
+@pytest.mark.asyncio
+async def test_exit_transaction_mnemonic_multiple() -> None:
     # Prepare folder
     my_folder_path = os.path.join(os.getcwd(), 'TESTING_TEMP_FOLDER')
     clean_exit_transaction_folder(my_folder_path)
     if not os.path.exists(my_folder_path):
         os.mkdir(my_folder_path)
 
-    runner = CliRunner()
-    inputs = []
-    data = '\n'.join(inputs)
-    arguments = [
+    if os.name == 'nt':  # Windows
+        run_script_cmd = 'sh deposit.sh'
+    else:  # Mac or Linux
+        run_script_cmd = './deposit.sh'
+
+    install_cmd = run_script_cmd + ' install'
+    proc = await asyncio.create_subprocess_shell(
+        install_cmd,
+    )
+    await proc.wait()
+
+    cmd_args = [
+        run_script_cmd,
         '--language', 'english',
         '--non_interactive',
         'exit-transaction-mnemonic',
         '--output_folder', my_folder_path,
         '--chain', 'mainnet',
-        '--mnemonic', 'aban aban aban aban aban aban aban aban aban aban aban abou',
+        '--mnemonic', '"aban aban aban aban aban aban aban aban aban aban aban abou"',
         '--validator_start_index', '0',
-        '--validator_indices', '0 1 2 3',
+        '--validator_indices', '0,1,2,3',
         '--epoch', '1234',
     ]
-    result = runner.invoke(cli, arguments, input=data)
+    proc = await asyncio.create_subprocess_shell(
+        ' '.join(cmd_args),
+    )
+    await proc.wait()
 
-    assert result.exit_code == 0
+    assert proc.returncode == 0
 
     # Check files
     exit_transaction_folder_path = os.path.join(my_folder_path, DEFAULT_EXIT_TRANSACTION_FOLDER_NAME)

--- a/tests/test_cli/test_new_mnemonic.py
+++ b/tests/test_cli/test_new_mnemonic.py
@@ -409,6 +409,8 @@ async def test_script_bls_withdrawal() -> None:
 
     assert len(seed_phrase) > 0
 
+    await proc.wait()
+
     # Check files
     validator_keys_folder_path = os.path.join(my_folder_path, DEFAULT_VALIDATOR_KEYS_FOLDER_NAME)
     _, _, key_files = next(os.walk(validator_keys_folder_path))
@@ -494,6 +496,8 @@ async def test_script_abbreviated_mnemonic() -> None:
                 proc.stdin.write(b'\n')
 
     assert len(seed_phrase) > 0
+
+    await proc.wait()
 
     # Check files
     validator_keys_folder_path = os.path.join(my_folder_path, DEFAULT_VALIDATOR_KEYS_FOLDER_NAME)

